File: 31 Next Permutation/31 Next Permutation.py
"""
Implement next permutation, which rearranges numbers into the lexicographically next greater permutation of numbers.
If such an arrangement is not possible, it must rearrange it as the lowest possible order (i.e., sorted in ascending order).
The replacement must be in place and use only constant extra memory.

Example 1:

Input: nums = [1,2,3]
Output: [1,3,2]

Example 2:

Input: nums = [3,2,1]
Output: [1,2,3]

Example 3:

Input: nums = [1,1,5]
Output: [1,5,1]

Example 4:

Input: nums = [1]
Output: [1]
"""
class Solution:
    def nextPermutation(self, nums: List[int]) -> None:
        """
        Do not return anything, modify nums in-place instead.
        """
        n = len(nums)

        def swap(i, j):
            nums[i], nums[j] = nums[j], nums[i]

        def reverse(s):
            i = s
            j = n-1
            while i < j:
                swap(i, j)
                i += 1
                j -= 1

        i = n - 2
        while i >= 0 and nums[i] >= nums[i+1]:
            i -= 1

        if i >= 0:
            j = n - 1
            while nums[j] <= nums[i]:
                j -= 1

            swap(i, j)

        reverse(i+1)

        return nums

File: 31 Next Permutation/31 Next Permutation.go
func nextPermutation(nums []int)  {

  l:=len(nums)

    var i,index int
    for i=l-1;i>0;i--{
        if nums[i]>nums[i-1]{

            var j int
            for j=i;j<l;j++{
                if nums[j]<=nums[i-1]{
                    break
                }
            }
            j--
            index = i
            nums[j],nums[i-1]=nums[i-1],nums[j]

            break
        }
    }


    i,j:=index,l-1
        for i<j{
            nums[i],nums[j]=nums[j],nums[i]
            i++

            j--
        }

    return

}
